gel/orm/introspection.py

```diff
@@ -197,6 +199,32 @@ def _process_links(types, modules):
                         'target': target,
                     })
 
+    # Go over backlinks and resolve any name collisions using the type map.
+    for spec in types:
+        mod = spec["name"].rsplit('::', 1)[0]
+        sql_source = get_sql_name(spec["name"])
+
+        # Find collisions in backlink names
+        bk = collections.defaultdict(list)
+        for link in spec['backlinks']:
+            if link['name'].startswith('backlink_via_'):
+                bk[link['name']].append(link)
+
+        for bklinks in bk.values():
+            if len(bklinks) > 1:
+                # We have a collision, so each backlink in it must now be
+                # disambiguated.
+                for link in bklinks:
+                    origsrc = get_sql_name(link['target']['name'])
+                    lname = link['name']
+                    link['name'] = f'{lname}_from_{origsrc}'
+                    # Also update the original source of the link with the
+                    # special backlink name.
+                    source = type_map[link['target']['name']]
+                    fwname = lname.replace('backlink_via_', '', 1)
+                    link['fwname'] = fwname
+                    source['backlink_renames'][fwname] = link['name']
```

Comment:
Could we *always* generate the disambiguated backlink, and then only generate the short name when it is unambiguous, or does that not work?

Reply:
We mostly expect not too many collisions so I tend to treat that as a the default. Backlink collisions appear when everything is linked to everything, which is not necessarily a common schema since that creates a lot of links to maintain.
Also, detecting unambiguous backlinks is a little more annoying that the other way around. Collisions are shown by identical names, but unambiguous long names would have to be checked by looking up their corresponding forward link name, seems more roundabout.